packages/node/src/starknet/decodeCalldata.ts

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import { FELT, StarknetContractCall } from '@subql/types-starknet';
+
+export class DecodeCalldataError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'Decode Calldata Error';
+  }
+}
+
+/***
+ * Decode the calldata for a generic contract call, like INVOKE V0 type or L1 contract call
+ * @param contractAddress
+ * @param selector
+ * @param calldata
+ */
+export function decodeGenericCalldata(
+  contractAddress: FELT,
+  selector: FELT,
+  calldata: FELT[],
+): StarknetContractCall {
+  return {
+    to: contractAddress,
+    selector: selector,
+    calldata: calldata,
+  };
+}
+
+/***
+ * Decode the INVOKE v2 & v3 calldata in contract call
+ * This been rewritten from rust version https://github.com/9oelM/decode-starknet-calldata/blob/main/src/lib.rs
+ *
+ * @param calldata
+ */
+export function decodeInvokeCalldata(calldata: FELT[]): StarknetContractCall[] {
```

Comment:
What is the performance of this like? Is it something that could be done lazily or only after filter checks?

Reply:
Haven't test its performance yet. 
Now I adjusted the logic, it should only decode when filter is used, or when user parse the call.